### PR TITLE
Partial fix for issue with extending relation with a gap.

### PR DESCRIPTION
**Pull request by hlinnaka**
This should fix the sporadic regression test failures we've been seeing
lately with "no base img found" errors.

This fixes the common case, but one corner case is still not handled:
If a relation is extended across a segment boundary, leaving a gap block
in the segment preceding the segment containing the target block, the
preceding segment will not be padded with zeros correctly. This adds
a test case for that, but it's commented out.

See github issue https://github.com/zenithdb/zenith/issues/500

### CODE REVIEW

**Comment by hlinnaka** [comment]
This is a quick fix to make the sporadic regression failures go away. I've got a feeling that we need to re-think the way we track relation size, and how we treat these gaps again. Perhaps the check for whether a put-call extends the relation or not should be moved into layered_repository.rs, because InMemoryLayer::put_page_version function cannot really deal with cross-segment issues.

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 1]
```diff
@@ -334,6 +335,37 @@ impl InMemoryLayer {
...
                     lsn
                 );
+
+                // If we are extending the relation by more than one page, initialize the "gap"
+                // with zeros
+                //
+                // XXX: What if the caller initializes the gap with subsequent call with same LSN?
+                // I don't think that can happen currently, but that is highly dependent on how
+                // PostgreSQL writes its WAL records and there's guarantee on it. If it does
+                // happen, we would hit the "page version already exists" warning aboe on the
```
Comment (LizardWizzard) [inline pageserver/src/layered_repository/inmemory_layer.rs:345]: typo: aboe  -> about

**Review by LizardWizzard — APPROVED** [review]
Just to clarify. When we receive blkno which enlarges current segment past it's boundary, during conversion to segment tag we get tag off the next not yet existing segment right? And this case is one that fails with this changes, so this patch handles gaps in a single segment filling them with zeroes, correct?

**Comment by hlinnaka** [comment]
> Just to clarify. When we receive blkno which enlarges current segment past it's boundary, during conversion to segment tag we get tag off the next not yet existing segment right? And this case is one that fails with this changes, so this patch handles gaps in a single segment filling them with zeroes, correct?

That's right.

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 2]
```diff
@@ -334,6 +335,37 @@ impl InMemoryLayer {
                     newsize,
                     lsn
                 );
+
+                // If we are extending the relation by more than one page, initialize the "gap"
+                // with zeros
+                //
+                // XXX: What if the caller initializes the gap with subsequent call with same LSN?
+                // I don't think that can happen currently, but that is highly dependent on how
+                // PostgreSQL writes its WAL records and there's guarantee on it. If it does
```
Comment (patins) [inline pageserver/src/layered_repository/inmemory_layer.rs:344]: no guarantee?

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 3]
```diff
@@ -334,6 +335,37 @@ impl InMemoryLayer {
                     newsize,
                     lsn
                 );
+
+                // If we are extending the relation by more than one page, initialize the "gap"
+                // with zeros
+                //
+                // XXX: What if the caller initializes the gap with subsequent call with same LSN?
```
Comment (patins) [inline pageserver/src/layered_repository/inmemory_layer.rs:342]: just for my own understanding, what would we see in args for these two calls to `put_page_version`?

Comment (hlinnaka) [reply pageserver/src/layered_repository/inmemory_layer.rs:342]: Something like this:

(relation size is 9 at the start, i.e. blk number 8 is the last block)
put_page_version(10, Lsn(123), ...); // extends the relation by two blocks, fills block 9 with zeros
put_page_version(9, Lsn(123), ...); // writes real data to block 9 at the same LSN

Comment (patins) [reply pageserver/src/layered_repository/inmemory_layer.rs:367]: Ahh I see. 
Maybe we could store if the page had been initialized empty for the current LSN

**Review by patins — APPROVED** [review]
Looks good, just consider for loop comment

**Comment by hlinnaka** [comment]
Turned the while- into a for-loop, and pushed. Thanks!
